File: Ashutosh-Singh/BinarySearch/Minimum-Limit-of-Balls-in-a-Bag.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int minimumSize(vector<int>& nums, int maxOperations) {
        int start = 1, end = *max_element(nums.begin(), nums.end());
        int minPenalty = end;

        // Binary search on possible penalty range
        while (start <= end) {
            int penalty = start + (end - start) / 2;
            if (isPossible(nums, maxOperations, penalty)) {
                minPenalty = penalty;
                end = penalty - 1;
            } else {
                start = penalty + 1;
            }
        }

        return minPenalty;
    }

private:
    bool isPossible(vector<int>& nums, int maxOperations, int penalty) {
        int requiredOps = 0;
        for (int n : nums) {
            int x = n / penalty;
            if (n % penalty == 0) x--; // Avoid overestimating operations for exact divisions
            requiredOps += x;
            if (requiredOps > maxOperations) return false; // Early exit if operations exceed the limit
        }
        return requiredOps <= maxOperations;
    }
};

int main() {
    Solution solution;
    vector<int> nums = {9, 6, 7, 11}; // Example input
    int maxOperations = 5;           // Example maximum operations

    int result = solution.minimumSize(nums, maxOperations);
    cout << "Minimum Penalty: " << result << endl;

    return 0;
}

File: Anya_Seth/BinarySearch/Minimum Limit of Balls in a Bag.java
class Solution {
    public int minimumSize(int[] nums, int maxOperations){
        int min = 1;
        int max = 0;
        for(int i = 0 ; i < nums.length; i++){
            if(max < nums[i]){
                max = nums[i];
            }
        }
        while(min < max){
            int mid =  (min + max)/2;
            if(isPossible(mid,nums,maxOperations)){
                max = mid;
            }
            else{
                min = mid + 1;
            }
        }
        return min;
    }
    private boolean isPossible(int maxBallsInBag, int[] nums, int maxOperations){
        int totalOperations = 0;
        for(int num : nums){
            int operations = (int)Math.ceil(num/(double)maxBallsInBag) - 1;
            totalOperations += operations;
            if(totalOperations > maxOperations){
                return false;
            }
        }
        return true;
        }
    }
